extension/src/reboot.rs

```diff
@@ -0,0 +1,243 @@
+	{
+		type C = C;
+
+		fn execute<E: Ext<T = C>>(mut env: Environment<E, BufInBufOutState>) -> Result<RetVal> {
+			// Build call
+			let call = B::build(&mut env)?.into();
+			// Charge weight before dispatch
+			let charged_weight = env.charge_weight(call.get_dispatch_info().weight)?;
+			// Ensure call allowed
+			let mut origin: C::RuntimeOrigin =
+				RawOrigin::Signed(env.ext().address().clone()).into();
+			origin.add_filter(Filter::contains);
+
+			let (result, weight) = match call.dispatch(origin) {
+				Ok(info) => (Ok(()), info.actual_weight),
+				Err(err) => (Err(err.error), err.post_info.actual_weight),
+			};
+			// Adjust post-dispatch weight
+			if let Some(actual_weight) = weight {
+				env.adjust_weight(charged_weight, actual_weight);
+			}
+			result.map(|_| Converging(0))
+		}
+	}
+
+	impl<C, B, D, M: Matches, F> Matches for DispatchCall<C, B, D, M, F> {
+		fn matches(func_id: u16) -> bool {
+			M::matches(func_id)
+		}
+	}
+
+	// Function implementation for reading state.
+	pub struct ReadState<C, B, R, RR, M, F>(PhantomData<(C, B, R, RR, M, F)>);
+	impl<
+			C: Config,
+			B: Builder<C = C, Output = R>,
+			R: Decode + Into<RR>,
+			RR: RuntimeRead,
+			Matcher: Matches,
+			Filter: Contains<RR>,
+		> Function for ReadState<C, B, R, RR, Matcher, Filter>
+	{
+		type C = C;
+
+		fn execute<E: Ext<T = C>>(mut env: Environment<E, BufInBufOutState>) -> Result<RetVal> {
+			// Build call
+			let read = B::build(&mut env)?.into();
+
+			// Charge weight before read
+			env.charge_weight(read.weight())?;
+			// Ensure call allowed
+			// TODO: use filtered error so easier to determine if it is decoding error or a filter blocking the call
+			ensure!(Filter::contains(&read), UNKNOWN_CALL_ERROR);
+			// TODO: check remaining parameters (allow_skip, weight per byte)
+			env.write(&read.read(), false, None)?;
+			Ok(Converging(0))
+		}
+	}
+
+	impl<C, B, R, RR, M: Matches, F> Matches for ReadState<C, B, R, RR, M, F> {
+		fn matches(func_id: u16) -> bool {
+			M::matches(func_id)
+		}
+	}
+
+	pub mod builders {
+		use super::*;
+
+		// Trait for building some output (call/read) from the environment.
+		pub trait Builder {
+			type C: Config;
+			type Output: Decode;
+			fn build<E: Ext<T = Self::C>>(
+				env: &mut Environment<E, BufInBufOutState>,
+			) -> Result<Self::Output>;
+		}
+
+		// Default implementation which charges appropriate weight before building the output.
+		pub struct DefaultBuilder<C, O>(PhantomData<(C, O)>);
+		impl<C: Config, Output: Decode> Builder for DefaultBuilder<C, Output> {
+			type C = C;
+			type Output = Output;
+
+			fn build<E: Ext<T = Self::C>>(
+				env: &mut Environment<E, BufInBufOutState>,
+			) -> Result<Self::Output> {
+				let contract_host_weight = <C as Config>::Schedule::get().host_fn_weights;
+				let len = env.in_len();
+				env.charge_weight(contract_host_weight.return_per_byte.saturating_mul(len.into()))?;
+				let params = env.read(len)?;
```

Comment:
This could potentially be called in `fn call` and `params` given as a parameter to `execute` and `build`.

Reply:
Seems very outdated based on latest push, let me know if still needs addressing in latest version.

Reply:
@Daanvdplas can this comment be resolved?

Reply:
yes, apologies, resolved